x-pack/legacy/plugins/siem/server/lib/detection_engine/routes/rules/utils.ts

```diff
@@ -224,3 +227,41 @@ export const getDuplicates = (lodashDict: Dictionary<number>): string[] => {
   }
   return [];
 };
+
+export const getTupleDuplicateErrorsAndUniqueRules = (
+  rules: PromiseFromStreams[],
+  isOverwrite: boolean
+): [BulkError[], PromiseFromStreams[]] => {
+  const { errors, rulesAcc } = rules.reduce(
+    (acc, parsedRule) => {
+      if (parsedRule instanceof Error) {
+        acc.rulesAcc.set(uuid.v4(), parsedRule);
+      } else {
+        const { rule_id: ruleId } = parsedRule;
+        if (ruleId != null) {
+          if (acc.rulesAcc.has(ruleId) && !isOverwrite) {
+            acc.errors.set(
+              uuid.v4(),
+              createBulkErrorObject({
+                ruleId,
+                statusCode: 400,
+                message: `More than one rule with rule-id: "${ruleId}" found`,
+              })
+            );
+          }
+          acc.rulesAcc.set(ruleId, parsedRule);
+        } else {
+          acc.rulesAcc.set(uuid.v4(), parsedRule);
+        }
+      }
+
+      return acc;
+    }, // using map (preserves ordering)
+    {
+      errors: new Map<string, BulkError>(),
+      rulesAcc: new Map<string, PromiseFromStreams>(),
```

Comment:
Really awesome these are typed now and you are passing this through the reducer.

This function is a bit tricky to read and reason through (some functions are always like that) but with the tests and the way it is written I think I understand it.